libparsec/crates/client/src/workspace/store/resolve_path.rs

```diff
@@ -727,69 +795,273 @@ fn cache_only_retrieve_path_from_id(
+/// - The entry itself will only be fetched from local data.
+///
+/// This is because this method is expected to be used to synchronize this very
+/// entry from/to the server.
+pub(crate) async fn retrieve_path_from_id_and_lock_for_update(
     store: &super::WorkspaceStore,
     entry_id: VlobID,
-) -> Result<(ArcLocalChildManifest, FsPath, PathConfinementPoint), ResolvePathError> {
+    wait: bool,
+) -> Result<
+    (RetrievePathFromIDEntry, ManifestUpdateLockGuard),
+    RetrievePathFromIdAndLockForUpdateError,
+> {
+    // In this function, not all cache misses are going to be successfully populated:
+    // - If the entry is not present in local obviously.
+    // - If a parent in the path has an invalid ID.
+    //
+    // However event in those cases we need to take the update lock guard for the entry.
+    //
+    // Hence this `last_not_found_during_populate` variable that gets provided to
+    // the cache-only retrieval function so that it knows if it should ask for a
+    // populate or just abort and return a missing/unreachable outcome.
+    let mut last_not_found_during_populate = None;
     loop {
         // Most of the time we should have each entry in the path already in the cache,
         // so we want to lock the cache once and only release it in the unlikely case
         // we need to fetch from the local storage or server.
         let cache_only_outcome = {
             let mut cache = store.current_view_cache.lock().expect("Mutex is poisoned");
-            cache_only_retrieve_path_from_id(&mut cache, entry_id)
+            cache_only_retrieve_path_from_id(
+                &mut cache,
+                entry_id,
+                true,
+                last_not_found_during_populate,
+            )
         };
         match cache_only_outcome {
-            CacheOnlyRetrievalOutcome::Done((entry_manifest, path, confinement_point)) => {
-                return Ok((entry_manifest, path, confinement_point))
+            CacheOnlyPathRetrievalOutcome::Done {
+                entry,
+                maybe_update_lock_guard,
+            } => {
+                let lock_guard = maybe_update_lock_guard.expect("always present");
+                return Ok((entry, lock_guard));
             }
-            CacheOnlyRetrievalOutcome::EntryNotFound => {
-                return Err(ResolvePathError::EntryNotFound)
+
+            // We got a cache miss (entry is missing)
+            CacheOnlyPathRetrievalOutcome::NeedPopulateCache(cache_miss_entry_id)
+                if cache_miss_entry_id == entry_id =>
+            {
+                match populate_cache_from_local_storage(store, cache_miss_entry_id).await {
+                    // Happy case :)
+                    Ok(_) => (),
+                    Err(PopulateCacheFromLocalStorageError::EntryNotFound) => {
+                        last_not_found_during_populate = Some(cache_miss_entry_id);
```

Comment:
Isn't it more readable to return a `RetrievePathFromIDEntry::Missing` here? 

Or maybe you need re-enter `cache_only_retrieve_path_from_id` for some lock handling reason?

Reply:
yes that's the thing: `cache_only_retrieve_path_from_id` is responsible to resolve the path and provide the update lock in one atomic step